### ISSUE TITLE
Add ability to match on nodegroups to the compound matcher

### ISSUE THREAD

**Issue by corywright** [saltstack/salt#6602]
As discussed in #4600 it is currently not possible to have a compound match on nodegroups.  

It would be very useful to be able to group nodes into environments and roles, then target them like this:

    # salt -C 'N@env0 and N@web' state.highstate
    # salt -C 'N@api or N@web' state.highstate

Per #4600, there were plans to get this into 0.15, then 0.16, then 0.17, but it doesn't look like it's been prioritized yet.

**Comment by plastikos** [comment]
See this PR: https://github.com/saltstack/salt/pull/22491

**Comment by rallytime** [comment]
@corywright, or anyone else on this thread, please give the pull request above a try and let us know if we can close out this request. 

Thanks @plastikos!

**Comment by rallytime** [comment]
Since we haven't heard back here, I am going to close this as I think the pull request above has implemented this functionality nicely. If anyone disagrees or more work should be done here, please leave a comment and we'll be glad to re-open.

**Comment by The-Loeki** [comment]
I respectfully disagree; the aforementioned fix only applies to nodegroup matching **within nodegroups**; `salt -C 'E@myhosts* or N@mygroup'` still can't work for example

**Comment by rallytime** [comment]
@The-Loeki No problem. 

@plastikos If you've got a moment, would you mind taking a look at this use case?

**Comment by avoid3d** [comment]
Any progress on this issue? How hard would you estimate this is to fix, never contributed to Saltstack, is this noobie compatible?

**Comment by jasrusable** [comment]
+1 for this issue.

**Comment by dreadworks** [comment]
+1, as it would remove the necessity to define millions of nodegroups to reduce matches:

`'N@mygroup and not special.*'`

**Comment by DanyC97** [comment]
+1

**Comment by plastikos** [comment]
I believe this *does* work as @The-Loeki describes - I'll check it in the morning.

**Comment by timo-klarshift** [comment]
This does not work:

top.sls (pillar)
```yaml
'N@mygroup':
  - match: compount
  - pillar.foo
```
works using match: nodegroup does ...

Also this does not work: 
```sh
salt-call -C 'N@mygroup' test.ping
```
works using -N 'mygroup'

**Comment by basepi** [comment]
`salt-call` only executes on the minion you run it on. It does not take a targeting string. To use compound matching at all you need to be on the master and use the `salt` command, not `salt-call`.

Additionally, you have a typo in your pillar topfile. `compound`, not `compount`.

**Comment by The-Loeki** [comment]
```
[root@mgmt.ceph:~]$ salt -C 'N@mon' test.ping
No minions matched the target. No command was sent, no jid was assigned.
ERROR: No return received
[root@mgmt.ceph:~]$ salt -N 'mon' test.ping
mon2.ceph:
    True
mon1.ceph:
    True
mon3.ceph:
    True
```

**Comment by basepi** [comment]
@plastikos Could you look into this when you have a moment?

**Comment by ahammond** [comment]
+1 this is a blocker for @white-hat and me. We want to use a nodegroup to keep track of minions which should never be touched by salt except via manual intervention. For example, we want to have a reactor run the following orchestration:

```yaml
update_inventory:
  salt.state:
    - tgt: L@{{ reactor_data.update_hosts }} and not N@untouchables
    - tgt_type: compound
    - highstate: True
```

**Comment by blaketmiller** [comment]
+1 this is a blocker for me as well. I have node groups that are matched against 2 data centers and I'd like to be able to isolate on each data center like: `salt -C 'N@elasticsearch and G@datacenter:west' test.ping`

**Comment by ahammond** [comment]
Our workaround was to make a nodegroup for every reasonable compound. Which
works well enough for us.

**Comment by plastikos** [comment]
This *used* to work.  I'll look at it.

**Comment by corywright** [comment]
Any idea when we may see this implemented?

**Comment by The-Loeki** [comment]
Well it didn't make into 2017.7 :(

**Comment by sokratisg** [comment]
Since Salt API is run on salt-master, can we use nodegroup targeting via the API?
For example, will the following API call work as intended?
```
[{'client': 'local', 'expr_form': 'compound', 'tgt': nodegroup1, 'fun': 'test.ping'}]
```

If yes, thes I suppose same positive answer applies for salt runners such as orchestrate, right?
